packages/block-library/src/navigation-link/index.php

```diff
@@ -367,5 +387,42 @@ function register_block_core_navigation_link() {
 			'variations'      => array_merge( $built_ins, $variations ),
 		)
 	);
+
+	// Register actions for all post types and taxonomies that may not yet be registered.
+	// This needs to happen in this function, because otherwise any post types/taxonomies
+	// registered **before** this block will attempt to register on a block that does not yet exist.
+	add_action( 'registered_post_type', 'register_block_core_navigation_link_post_type_variation', 10, 2 );
+	add_action( 'registered_taxonomy', 'register_block_core_navigation_link_taxonomy_variation', 10, 3 );
 }
 add_action( 'init', 'register_block_core_navigation_link' );
+
+/**
+ * Register custom post type variations for navigation link on post type registration
+ * Handles all post types registered after the block is registered in register_navigation_link_post_type_variations
+ *
+ * @param string       $post_type The post type name passed from registered_post_type filter.
+ * @param WP_Post_Type $post_type_object The post type object passed from registered_post_type.
+ * @return void
+ */
+function register_block_core_navigation_link_post_type_variation( $post_type, $post_type_object ) {
+	if ( $post_type_object->show_in_nav_menus ) {
+		$variation = build_variation_for_navigation_link( $post_type_object, 'post-type' );
+		register_block_core_navigation_link_variation( $variation );
+	}
+}
+
+/**
+ * Register a custom taxonomy variation for navigation link on taxonomy registration
+ * Handles all taxonomies registered after the block is registered in register_navigation_link_post_type_variations
+ *
+ * @param string       $taxonomy Taxonomy name.
+ * @param array|string $object_type Name of the object type for the taxonomy object.
+ * @param array|string $args Optional args used in taxonomy registration.
+ * @return void
+ */
+function register_block_core_navigation_link_taxonomy_variation( $taxonomy, $object_type, $args ) {
```

Comment:
```suggestion
function register_block_core_navigation_link_taxonomy_variation( $taxonomy, $object_type, $args ) {
```

@ramonjd I'm wondering whether we need to prefix these with `gutenberg_` as per docs in https://github.com/WordPress/gutenberg/blob/trunk/lib/README.md?

Reply:
Thanks for the ping @getdave 

In general block files shouldn't use `gutenberg` prefixes. They're ported to WordPress Core mostly as-is via package sync.

There are some exceptions, but when `gutenberg_` functions are meant to run when the plugin is activated, hence we see that they're wrapped in `defined( 'IS_GUTENBERG_PLUGIN' )`.

I can add this to the lib docs too to avoid ambiguity. 

Cheers!

Reply:
> I can add this to the lib docs too to avoid ambiguity.

Done here: https://github.com/WordPress/gutenberg/pull/55402

Also it's worth noting there are block library-specific naming conventions outlined over in https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/README.md#naming-convention-for-php-functions

👍🏻

Reply:
Thank you Ramon. Not action require here now.